File: leetcode/Medium/0402-remove-k-digits/0402-remove-k-digits.py
"""
 Problem: 402. Remove K Digits
 Language: python
 Runtime: 53 ms (32.80%)
 Memory: 13.2 MB (99.52%)
 Tags: String, Stack, Greedy, Monotonic Stack
"""
class Solution(object):
    def removeKdigits(self, num, k):
        """
        :type num: str
        :type k: int
        :rtype: str
        """
        stack = []
        
        for digit in num:
            while stack and k > 0 and stack[-1] > digit:
                stack.pop()
                k -= 1
            stack.append(digit)
        
        stack = stack[:-k] if k > 0 else stack
        
        result = ''.join(stack).lstrip('0')
        
        return result if result else '0'                
        

File: leetcode/Medium/0402-remove-k-digits/0402-remove-k-digits.cpp
/*
 Problem: 402. Remove K Digits
 Language: cpp
 Runtime: 0 ms (100.00%)
 Memory: 11.1 MB (82.55%)
 Tags: String, Stack, Greedy, Monotonic Stack
*/
class Solution {
public:
    string removeKdigits(string num, int k) {
        string st;

        for (const auto& c : num) {
            while (!st.empty() && k > 0 && st.back() > c) {
                st.pop_back();
                --k;
            }
            st.push_back(c);
        }

        while (!st.empty() && k > 0) {
            st.pop_back();
            --k;
        }

        int i = 0;
        while (i < st.size() && st[i] == '0') ++i;

        string res = st.substr(i);
        return res.empty() ? "0" : res;
    }
};